File: beam-api/app/schemas/analysis.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.core.config import settings


class AnalysisStatus(str, Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


# Schema-level safety net only; the route enforces the configured limit with 413.
_SCHEMA_TEXT_HARD_MAX = settings.MAX_ANALYSIS_TEXT_CHARS * 4


class AnalysisCreate(BaseModel):
    text: str = Field(min_length=1, max_length=_SCHEMA_TEXT_HARD_MAX)
    source_type: str = Field(default="text", min_length=1, max_length=100)
    title: str | None = Field(default=None, max_length=255)

    @field_validator("text")
    @classmethod
    def text_must_not_be_blank(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("text must not be empty or whitespace-only")
        return stripped

    @field_validator("source_type", "title")
    @classmethod
    def strip_optional(cls, value: str | None) -> str | None:
        if value is None:
            return None
        return value.strip() or None


class AnalysisCreatedResponse(BaseModel):
    session_id: UUID
    status: AnalysisStatus


class EmotionDistributionItem(BaseModel):
    label: str
    score: float


class PredictionPayload(BaseModel):
    """Model-derived prediction. All fields are null until a real model runs."""

    model_config = ConfigDict(from_attributes=True)

    primary_emotion: str | None = None
    confidence: float | None = None
    emotion_distribution: list[EmotionDistributionItem] | None = None
    inference_time_ms: int | None = None


class BehaviorMetricsPayload(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    positivity_score: float | None = None
    negativity_score: float | None = None
    engagement_score: float | None = None
    linguistic_complexity: float | None = None
    emotional_variance: float | None = None
    posting_frequency: float | None = None


class ExplanationPayload(BaseModel):
    method: str | None = None
    summary: str | None = None
    important_keywords: list[str] | None = None


class ModelInfo(BaseModel):
    model_name: str | None = None
    model_version: str | None = None
    deployed: bool = False
    note: str | None = None


class AnalysisInputMetadata(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    char_count: int
    word_count: int
    raw_text: str
    created_at: datetime


class AnalysisSessionSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    session_id: UUID
    status: AnalysisStatus
    source_type: str
    title: str | None
    created_at: datetime
    completed_at: datetime | None
    primary_emotion: str | None = None
    confidence: float | None = None
    model_name: str | None = None


class AnalysisDetailResponse(AnalysisSessionSummary):
    input: AnalysisInputMetadata
    prediction: PredictionPayload | None = None
    behavior_metrics: BehaviorMetricsPayload | None = None
    explanation: ExplanationPayload | None = None
    model_info: ModelInfo


class AnalysisListResponse(BaseModel):
    items: list[AnalysisSessionSummary]
    total: int
    page: int
    page_size: int
    pages: int

File: beam-web/src/types/analysis.ts
export type AnalysisStatus = 'pending' | 'processing' | 'completed' | 'failed'

export type SourceType = 'text' | 'discussion' | 'social_feed' | 'review'

export type AnalysisCreateRequest = {
  text: string
  source_type: SourceType
  title?: string | null
}

export type AnalysisCreatedResponse = {
  session_id: string
  status: AnalysisStatus
}

export type EmotionDistributionItem = {
  label: string
  score: number
}

/** Model-derived prediction. Fields are null until a real model is deployed. */
export type PredictionPayload = {
  primary_emotion: string | null
  confidence: number | null
  emotion_distribution: EmotionDistributionItem[] | null
  inference_time_ms: number | null
}

export type BehaviorMetricsPayload = {
  positivity_score: number | null
  negativity_score: number | null
  engagement_score: number | null
  linguistic_complexity: number | null
  emotional_variance: number | null
  posting_frequency: number | null
}

export type ExplanationPayload = {
  method: string | null
  summary: string | null
  important_keywords: string[] | null
}

export type ModelInfoPayload = {
  model_name: string | null
  model_version: string | null
  deployed: boolean
  note: string | null
}

export type AnalysisInputMetadata = {
  char_count: number
  word_count: number
  raw_text: string
  created_at: string
}

export type AnalysisSessionSummary = {
  session_id: string
  status: AnalysisStatus
  source_type: string
  title: string | null
  created_at: string
  completed_at: string | null
  primary_emotion: string | null
  confidence: number | null
  model_name: string | null
}

export type AnalysisDetail = AnalysisSessionSummary & {
  input: AnalysisInputMetadata
  prediction: PredictionPayload | null
  behavior_metrics: BehaviorMetricsPayload | null
  explanation: ExplanationPayload | null
  model_info: ModelInfoPayload
}

export type AnalysisSort = 'created_desc' | 'created_asc'

export type AnalysisListParams = {
  page?: number | undefined
  page_size?: number | undefined
  search?: string | undefined
  status?: AnalysisStatus | '' | undefined
  sort?: AnalysisSort | undefined
}

export type AnalysisListResponse = {
  items: AnalysisSessionSummary[]
  total: number
  page: number
  page_size: number
  pages: number
}
